example/sonivoxrender.1
.\" Automatically generated by Pandoc 2.14.0.3
.\"
.TH "SONIVOXRENDER" "1" "February 14, 2026" "sonivox 4.0.1.0" "Sonivox MIDI File Renderer"
.hy
.SH NAME
.PP
\f[B]sonivoxrender\f[R] \[em] Render MIDI song files into raw PCM audio
.SH SYNOPSIS
.PP
\f[B]sonivoxrender\f[R] [\f[B]-h|--help\f[R]] [\f[B]-v|--version\f[R]]
[\f[B]-d|--dls\f[R] \f[I]soundfont\f[R]] [\f[B]-r|--reverb\f[R]
\f[I]0..4\f[R]] [\f[B]-w|--wet\f[R] \f[I]0..32767\f[R]]
[\f[B]-n|--dry\f[R] \f[I]0..32767\f[R]] [\f[B]-c|--chorus\f[R]
\f[I]0..4\f[R]] [\f[B]-l|--level\f[R] \f[I]0..32767\f[R]]
[\f[B]-g|--gain\f[R] \f[I]0..196\f[R]] [\f[B]-V|--Verbosity\f[R]
\f[I]0..5\f[R]] [\f[B]-R|--reverb-post-mix\f[R]]
[\f[B]-C|--chorus-post-mix\f[R]] [\f[B]-s|--sndlib\f[R] \f[I]1..3\f[R]]
\f[I]song_file\f[R]
.SH DESCRIPTION
.PP
This program is a MIDI file renderer based on the sonivox synthesizer
library.
It reads .MID (Standard MIDI Files), .RMI (RIFF-RMID), and .XMF
(Extensible Music Files) file formats, and writes an audio stream to the
standard output as raw 16 bit stereo PCM samples.
.SS Options
.TP
-h, --help
Prints brief usage information.
.TP
-v, --version
Prints the version numbers.
.TP
-d, --dls \f[I]soundfont\f[R]
Optional DLS or SF2 soundfont file name.
If not provided, it uses an internal embedded soundfont.
.TP
-r, --reverb \f[I]reverb_preset\f[R]
Reverb preset between 0 and 4: 0=no, 1=large hall, 2=hall, 3=chamber,
4=room.
.TP
-w, --wet \f[I]reverb_wet\f[R]
Reverb wet level between 0 and 32767.
.TP
-n, --dry \f[I]reverb_dry\f[R]
Reverb dry level between 0 and 32767.
.TP
-c, --chorus \f[I]chorus_preset\f[R]
Chorus preset between 0 and 4: 0=no, 1..4=presets.
.TP
-l, --level \f[I]chorus_level\f[R]
Chorus level between 0 and 32767.
.TP
-g, --gain \f[I]master_gain\f[R]
Master gain between 0 and 196, default is 100 (+0dB).
The number is relative to 100, in 1dB increments, e.g.\ 120 = +20dB, 80
= -20dB.
.TP
-V, --Verbosity \f[I]verbosity\f[R]
Verbosity level between 0 and 5, where 0=no, 1..5=severity levels.
.TP
-R, --reverb-post-mix
Ignore CC91 reverb send level.
The reverb effect will apply to mixed output audio, which is the old
behavior.
.TP
-C, --chorus-post-mix
Ignore CC93 chorus send level.
See also \f[B]--reverb-post-mix\f[R].
.TP
-s, --sndlib \f[I]index\f[R]
EAS sound library to use:
.RS
.IP \[bu] 2
1: wt_200k_G (default) - WT-only bank, used in Android devices.
Also named \[lq]Common\[rq].
Support 22050 Hz and 44100 Hz sample rates, 8-bit and 16-bit samples.
.IP \[bu] 2
2: GMdblib-3 - FM-only bank.
Support all sample rates.
Sample bit depth does not matter here.
.IP \[bu] 2
3: hybrid_22khz_mcu - Hybrid bank.
Use WT synth for drums and FM for melodic instruments.
This bank is a combination of \f[C]GMdblib-3\f[R] and
\f[C]Sonic_20Khz_Drums\f[R].
Support 22050 Hz sample rate and 8-bit samples only.
.PP
\f[B]Note:\f[R] This option does not affect DLS/SF2.
They will always use the DLS synth engine.
If the selected sound library is not compatible with the build
configuration, the program will fail with an error message.
.RE
.SS Arguments
.TP
\f[I]song_file\f[R]
Input [\f[B]MID|RMI|XMF\f[R]] file name.
.SH EXAMPLES
.PP
The following examples assume the default option USE_44KHZ=ON, which
means an output sample rate = 44100 Hz.
.PP
Example 1: Render a MIDI file and save the rendered audio as a raw audio
file:
.IP
.nf
\f[C]
$ sonivoxrender ants.mid > ants.pcm
\f[R]
.fi
.PP
Example 2: pipe the rendered audio thru the Linux ALSA \f[B]aplay\f[R]
utility:
.IP
.nf
\f[C]
$ sonivoxrender ants.mid | aplay -c 2 -f S16_LE -r 44100
\f[R]
.fi
.PP
is equivalent to:
.IP
.nf
\f[C]
$ sonivoxrender ants.mid | aplay -f cd
\f[R]
.fi
.PP
Example 3: pipe the rendered audio thru the \f[B]lame\f[R] utility
creating a MP3 file:
.IP
.nf
\f[C]
$ sonivoxrender ants.mid | lame -r -s 44100 - ants.mp3
\f[R]
.fi
.PP
Example 4: pipe the rendered audio thru the \f[B]sox\f[R] utility
creating a WAV file:
.IP
.nf
\f[C]
$ sonivoxrender ants.mid | sox -t s16 -c 2 -r 44100 - ants.wav
\f[R]
.fi
.PP
Example 5: pipe the rendered audio thru the PulseAudio\[cq]s
\f[B]pacat\f[R] utility:
.IP
.nf
\f[C]
$ sonivoxrender ants.mid | pacat
\f[R]
.fi
.PP
Example 6: pipe the rendered audio thru the PipeWire\[cq]s
\f[B]pw-play\f[R] utility:
.IP
.nf
\f[C]
$ sonivoxrender ants.mid | pw-play --rate 44100 -
\f[R]
.fi
.PP
Example 7: pipe the rendered audio thru the FFmpeg\[cq]s
\f[B]ffplay\f[R] utility:
.IP
.nf
\f[C]
$ sonivoxrender ants.mid | ffplay -i - -f s16le -ar 44.1k -ac 2 -nodisp -autoexit -loglevel quiet
\f[R]
.fi
.PP
Example 8: pipe the rendered audio thru the \f[B]mpv\f[R] media player:
.IP
.nf
\f[C]
$ sonivoxrender ants.mid | mpv --demuxer=rawaudio -demuxer-rawaudio-format=s16le --demuxer-rawaudio-rate=44100 --demuxer-rawaudio-channels=2 --no-video -
\f[R]
.fi
.SH BUGS
.PP
See Tickets at GitHub <https://github.com/EmbeddedSynth/sonivox/issues/>
.SH LICENSE AND COPYRIGHT
.PP
Licensed under the Apache License, Version 2.0
.PP
Copyright (c) 2022-2026 Pedro L\['o]pez-Cabanillas and contributors
.SH AUTHORS
Pedro L\['o]pez-Cabanillas <plcl@users.sf.net>.


example/sonivoxrender.c
/*
 * Copyright (c) 2022-2026 Pedro López-Cabanillas
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <ctype.h>
#include <errno.h>
#include <fcntl.h>
#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <eas.h>
#include <eas_reverb.h>
#include <eas_chorus.h>
#include <eas_report.h>

#if defined(_MSC_VER)
#include <io.h>
#include <malloc.h>
#define alloca _alloca
#endif


const char *dls_path = NULL;
const char *sndlib_name = NULL;
EAS_I32 playback_gain = 100;
EAS_I32 reverb_type = 0;
EAS_I32 reverb_wet = 32767;
EAS_I32 reverb_dry = 0;
EAS_BOOL reverb_override = EAS_FALSE;
EAS_I32 chorus_type = 0;
EAS_I32 chorus_level = 32767;
EAS_BOOL chorus_override = EAS_FALSE;
EAS_DATA_HANDLE mEASDataHandle = NULL;
int verbosity =
#ifdef NDEBUG
    _EAS_SEVERITY_WARNING;
#else
    _EAS_SEVERITY_INFO;
#endif
const S_EAS_LIB_CONFIG *mEASConfig = NULL;
char sLibrary_version[16];

void initLibraryVersion()
{
    memset(sLibrary_version, 0, sizeof(sLibrary_version));
    mEASConfig = EAS_Config();
    if (mEASConfig == NULL) {
        fprintf(stderr, "Failed to get the library configuration\n");
    } else {
        unsigned char v1, v2, v3, v4;
        v1 = (mEASConfig->libVersion >> 24) & 0xff;
        v2 = (mEASConfig->libVersion >> 16) & 0xff;
        v3 = (mEASConfig->libVersion >> 8) & 0xff;
        v4 = mEASConfig->libVersion & 0xff;
        snprintf(sLibrary_version, sizeof(sLibrary_version), "%u.%u.%u.%u", v1, v2, v3, v4);
    }
}

void shutdownLibrary(void)
{
    if (mEASDataHandle) {
        EAS_RESULT result = EAS_Shutdown(mEASDataHandle);
        if (result != EAS_SUCCESS) {
            fprintf(stderr, "Failed to deallocate the resources for synthesizer library\n");
        }
    }
}

int initializeLibrary(void)
{
    int ok = EXIT_SUCCESS;

#if defined(__MINGW32__) 
    setmode(fileno(stdout), O_BINARY);
#elif defined(_MSC_VER)
    _setmode(_fileno(stdout), _O_BINARY);
#endif

    EAS_SetDebugFile(stderr, 1);
    EAS_SetDebugLevel(verbosity);

    EAS_RESULT result = EAS_Init(&mEASDataHandle);
    if (result != EAS_SUCCESS) {
        fprintf(stderr, "Failed to initialize synthesizer library\n");
        ok = EXIT_FAILURE;
        return ok;
    }

    if (mEASDataHandle == NULL) {
        fprintf(stderr, "Failed to initialize EAS data handle\n");
        ok = EXIT_FAILURE;
        return ok;
    }

    if (sndlib_name == NULL) {
        sndlib_name = EAS_GetDefaultSoundLibrary(EAS_SNDLIB_WT);
        if (sndlib_name == NULL) {
            fprintf(stderr, "Failed to get default sound library name\n");
            ok = EXIT_FAILURE;
            return ok;
        }
    }
    EAS_Report(_EAS_SEVERITY_DETAIL, "Using sound library: %s\n", sndlib_name);
    result = EAS_SetSoundLibrary(mEASDataHandle, NULL, EAS_GetSoundLibrary(mEASDataHandle, sndlib_name));
    if (result != EAS_SUCCESS) {
        fprintf(stderr, "Failed to set sound library: %s\n", sndlib_name);
        ok = EXIT_FAILURE;
        return ok;
    }

    if (dls_path != NULL) {
        EAS_FILE mDLSFile;
        memset(&mDLSFile, 0, sizeof(EAS_FILE));

        mDLSFile.handle = fopen(dls_path, "rb");
        if (mDLSFile.handle == NULL) {
            fprintf(stderr, "Failed to open %s. error: %s\n", dls_path, strerror(errno));
            ok = EXIT_FAILURE;
            goto cleanup;
        }

        result = EAS_LoadDLSCollection(mEASDataHandle, NULL, &mDLSFile);
        fclose(mDLSFile.handle);
        if (result != EAS_SUCCESS) {
            fprintf(stderr, "Failed to load DLS file\n");
            ok = EXIT_FAILURE;
            goto cleanup;
        }
    }

    result = EAS_SetVolume(mEASDataHandle, NULL, playback_gain);
    if (result != EAS_SUCCESS) {
        fprintf(stderr, "Failed to set master gain\n");
        ok = EXIT_FAILURE;
        goto cleanup;
    }

    result = EAS_SetParameter(mEASDataHandle, EAS_MODULE_REVERB, EAS_PARAM_REVERB_OVERRIDE_CC, reverb_override);
    if (result != EAS_SUCCESS) {
        fprintf(stderr, "Failed to enable reverb override\n");
        ok = EXIT_FAILURE;
        goto cleanup;
    }

    EAS_BOOL reverb_bypass = EAS_TRUE;
    EAS_I32 reverb_preset = reverb_type - 1;
    if ( reverb_preset >= EAS_PARAM_REVERB_LARGE_HALL && reverb_preset <= EAS_PARAM_REVERB_ROOM ) {
        reverb_bypass = EAS_FALSE;
        result = EAS_SetParameter(mEASDataHandle, EAS_MODULE_REVERB, EAS_PARAM_REVERB_PRESET, reverb_preset);
        if (result != EAS_SUCCESS) {
            fprintf(stderr, "Failed to set reverb preset");
            ok = EXIT_FAILURE;
            goto cleanup;
        }
        result = EAS_SetParameter(mEASDataHandle,
                                EAS_MODULE_REVERB,
                                EAS_PARAM_REVERB_WET,
                                reverb_wet);
        if (result != EAS_SUCCESS) {
            fprintf(stderr, "Failed to set reverb wet amount");
            ok = EXIT_FAILURE;
            goto cleanup;
        }

        result = EAS_SetParameter(mEASDataHandle, EAS_MODULE_REVERB, EAS_PARAM_REVERB_DRY, reverb_dry);
        if (result != EAS_SUCCESS) {
            fprintf(stderr, "Failed to set reverb dry amount");
            ok = EXIT_FAILURE;
            goto cleanup;
        }
    }
    result = EAS_SetParameter(mEASDataHandle, EAS_MODULE_REVERB, EAS_PARAM_REVERB_BYPASS, reverb_bypass);
    if (result != EAS_SUCCESS) {
        fprintf(stderr, "Failed to set reverb bypass");
        ok = EXIT_FAILURE;
        goto cleanup;
    }

    result = EAS_SetParameter(mEASDataHandle, EAS_MODULE_CHORUS, EAS_PARAM_CHORUS_OVERRIDE_CC, chorus_override);
    if (result != EAS_SUCCESS) {
        fprintf(stderr, "Failed to enable chorus override\n");
        ok = EXIT_FAILURE;
        goto cleanup;
    }

    EAS_BOOL chorus_bypass = EAS_TRUE;
    EAS_I32 chorus_preset = chorus_type - 1;
    if ( chorus_preset >= EAS_PARAM_CHORUS_PRESET1 && chorus_preset <= EAS_PARAM_CHORUS_PRESET4 ) {
        chorus_bypass = EAS_FALSE;
        result = EAS_SetParameter(mEASDataHandle, EAS_MODULE_CHORUS, EAS_PARAM_CHORUS_PRESET, chorus_preset);
        if (result != EAS_SUCCESS) {
            fprintf(stderr, "Failed to set chorus preset");
            ok = EXIT_FAILURE;
            goto cleanup;
        }

        result = EAS_SetParameter(mEASDataHandle, EAS_MODULE_CHORUS, EAS_PARAM_CHORUS_LEVEL, chorus_level);
        if (result != EAS_SUCCESS) {
            fprintf(stderr, "Failed to set chorus level");
            ok = EXIT_FAILURE;
            goto cleanup;
        }
    }

    result = EAS_SetParameter(mEASDataHandle, EAS_MODULE_CHORUS, EAS_PARAM_CHORUS_BYPASS, chorus_bypass);
    if (result != EAS_SUCCESS) {
        fprintf(stderr, "Failed to set chorus bypass");
        ok = EXIT_FAILURE;
        goto cleanup;
    }

    return ok;

cleanup:
    shutdownLibrary();

    return ok;
}

int renderFile(const char *fileName)
{
    EAS_HANDLE mEASStreamHandle = NULL;
    EAS_FILE mEasFile;
    EAS_PCM *mAudioBuffer = NULL;
    EAS_I32 mPCMBufferSize = 0;

    int ok = EXIT_SUCCESS;

    memset(&mEasFile, 0, sizeof(EAS_FILE));

    mEasFile.handle = fopen(fileName, "rb");
    if (mEasFile.handle == NULL) {
        fprintf(stderr, "Failed to open %s. error: %s\n", fileName, strerror(errno));
        ok = EXIT_FAILURE;
        return ok;
    }

    EAS_RESULT result = EAS_OpenFile(mEASDataHandle, &mEasFile, &mEASStreamHandle);
    if (result != EAS_SUCCESS) {
        fprintf(stderr, "Failed to open file\n");
        ok = EXIT_FAILURE;
        goto cleanup;
    }

    if(mEASStreamHandle == NULL) {
        fprintf(stderr, "Failed to initialize EAS stream handle\n");
        ok = EXIT_FAILURE;
        goto cleanup;
    }

    result = EAS_Prepare(mEASDataHandle, mEASStreamHandle);
    if (result != EAS_SUCCESS) {
        fprintf(stderr, "Failed to prepare EAS data and stream handles\n");
        ok = EXIT_FAILURE;
        goto cleanup;
    }

	EAS_I32 playLength = 0;
	result = EAS_ParseMetaData(mEASDataHandle, mEASStreamHandle, &playLength);
	if (result != EAS_SUCCESS) {
        fprintf(stderr, "Failed to parse MIDI file metadata\n");
        ok = EXIT_FAILURE;
        goto cleanup;
    }

	if (playLength == 0) {
		fprintf(stderr, "MIDI file time length returned 0\n");
        ok = EXIT_FAILURE;
        goto cleanup;
	}

    mEASConfig = EAS_Config();
    if (mEASConfig == NULL) {
        fprintf(stderr, "Failed to get the library configuration\n");
        ok = EXIT_FAILURE;
        goto cleanup;
    }

    mPCMBufferSize = sizeof(EAS_PCM) * mEASConfig->mixBufferSize * mEASConfig->numChannels;
    mAudioBuffer = alloca(mPCMBufferSize);
    if (mAudioBuffer == NULL) {
        fprintf(stderr, "Failed to allocate memory of size: %d", mPCMBufferSize);
        ok = EXIT_FAILURE;
        goto cleanup;
    }

    while (1) {
        EAS_STATE state;
        EAS_I32 count = -1;

        result = EAS_State(mEASDataHandle, mEASStreamHandle, &state);
        if (result != EAS_SUCCESS) {
            fprintf(stderr, "Failed to get EAS State\n");
            ok = EXIT_FAILURE;
            break;
        }

        if (state == EAS_STATE_STOPPED || state == EAS_STATE_ERROR) {
            break; /* playback is complete */
        }

        result = EAS_Render(mEASDataHandle, mAudioBuffer, mEASConfig->mixBufferSize, &count);
        if (result != EAS_SUCCESS) {
            fprintf(stderr, "Failed to render audio\n");
            ok = EXIT_FAILURE;
            break;
        }

        if (count != mEASConfig->mixBufferSize) {
            fprintf(stderr, "Only %d out of %d frames rendered\n", count, mEASConfig->mixBufferSize);
            ok = EXIT_FAILURE;
            break;
        }

        fwrite(mAudioBuffer, sizeof(EAS_PCM), mEASConfig->mixBufferSize * mEASConfig->numChannels, stdout);
        fflush(stdout);
    }

cleanup:
    if (mEASStreamHandle) {
        result = EAS_CloseFile(mEASDataHandle, mEASStreamHandle);
        if (result != EAS_SUCCESS) {
            fprintf(stderr, "Failed to close audio file/stream\n");
            ok = EXIT_FAILURE;
        }
    }

    if (mEasFile.handle != NULL) {
        fclose(mEasFile.handle);
    }
    return ok;
}

int main (int argc, char **argv)
{
    int ok = EXIT_SUCCESS;
    int index, c;

    initLibraryVersion();

    opterr = 0;
    int option_index = 0;

    static struct option long_options[] = {{"help", no_argument, 0, 'h'},
                                           {"version", no_argument, 0, 'v'},
                                           {"dls", required_argument, 0, 'd'},
                                           {"reverb", required_argument, 0, 'r'},
                                           {"wet", required_argument, 0, 'w'},
                                           {"dry", required_argument, 0, 'n'},
                                           {"chorus", required_argument, 0, 'c'},
                                           {"level", required_argument, 0, 'l'},
                                           {"gain", required_argument, 0, 'g'},
                                           {"Verbosity", required_argument, 0, 'V'},
                                           {"reverb-post-mix", no_argument, 0, 'R'},
                                           {"chorus-post-mix", no_argument, 0, 'C'},
                                           {"sndlib", required_argument, 0, 's'},
                                           {0, 0, 0, 0}};

    while (1) {
        c = getopt_long(argc, argv, "hvd:r:w:n:c:l:g:V:RCs:", long_options, &option_index);

        if (c == -1) {
            break;
        }

        switch (c) {
        case 'h':
            fprintf(
                stderr,
                "Usage: %s [-h|--help] [-v|--version] [-d|--dls soundfont] [-r|--reverb 0..4] "
                "[-w|--wet 0..32767] [-n|--dry 0..32767] "
                "[-c|--chorus 0..4] [-l|--level 0..32767] [-g|--gain 0..196] [-V|--Verbosity "
                "0..5] [-R|--reverb-post-mix] [-C|--chorus-post-mix] [-s|--sndlib 1..3] file.mid ...\n"
                "Render standard MIDI files into raw PCM audio.\n"
                "Options:\n"
                "\t-h, --help\t\tthis help message.\n"
                "\t-v, --version\t\tsonivox version.\n"
                "\t-d, --dls soundfont\tDLS or SF2 soundfont.\n"
                "\t-r, --reverb n\t\treverb preset: 0=no, 1=large hall, 2=hall, 3=chamber, "
                "4=room.\n"
                "\t-w, --wet n\t\treverb wet: 0..32767.\n"
                "\t-n, --dry n\t\treverb dry: 0..32767.\n"
                "\t-c, --chorus n\t\tchorus preset: 0=no, 1..4=presets.\n"
                "\t-l, --level n\t\tchorus level: 0..32767.\n"
                "\t-g, --gain n\t\tmaster gain: 0..196. 100 = +0dB.\n"
                "\t-V, --Verbosity n\tVerbosity: 0=no, 1=fatals, 2=errors, 3=warnings, 4=infos, "
                "5=details\n"
                "\t-R, --reverb-post-mix\tignore CC91 reverb send.\n"
                "\t-C, --chorus-post-mix\tignore CC93 chorus send.\n"
                "\t-s, --sndlib n\t\tsound engine library: 1=wt, 2=fm, 3=hybrid.\n",
                argv[0]);
            return EXIT_FAILURE;
        case 'v':
            fprintf(stderr, "version: %s\n", sLibrary_version);
            return EXIT_FAILURE;
        case 'd':
            dls_path = optarg;
            break;
        case 'R':
            reverb_override = EAS_TRUE;
            break;
        case 'C':
            chorus_override = EAS_TRUE;
            break;
        case 'r':
            reverb_type = atoi(optarg);
            if ((reverb_type < 0) || (reverb_type > 4)) {
                fprintf (stderr, "invalid reverb preset: %d\n", reverb_type);
                return EXIT_FAILURE;
            }
            break;
        case 'w':
            reverb_wet = atoi(optarg);
            if ((reverb_wet < 0) || (reverb_wet > 32767)) {
                fprintf (stderr, "invalid reverb wet: %d\n", reverb_wet);
                return EXIT_FAILURE;
            }
            break;
        case 'n':
            reverb_dry = atoi(optarg);
            if ((reverb_dry < 0) || (reverb_dry > 32767)) {
                fprintf (stderr, "invalid reverb dry: %d\n", reverb_dry);
                return EXIT_FAILURE;
            }
            break;
        case 'c':
            chorus_type = atoi(optarg);
            if ((chorus_type < 0) || (chorus_type > 4)) {
                fprintf (stderr, "invalid chorus preset: %d\n", chorus_type);
                return EXIT_FAILURE;
            }
            break;
        case 'l':
            chorus_level = atoi(optarg);
            if ((chorus_level < 0) || (chorus_level > 32767)) {
                fprintf (stderr, "invalid chorus level: %d\n", chorus_level);
                return EXIT_FAILURE;
            }
            break;
        case 'g':
            playback_gain = atoi(optarg);
            if ((playback_gain < 0) || (playback_gain > EAS_MAX_VOLUME)) {
                fprintf (stderr, "invalid playback gain: %d\n", playback_gain);
                return EXIT_FAILURE;
            }
            break;
        case 'V':
            verbosity = atoi(optarg);
            if ((verbosity < 0) || (verbosity > 5)) {
                fprintf(stderr, "invalid verbosity level: %d\n", verbosity);
                return EXIT_FAILURE;
            }
            break;
        case 's': {
            int sndlib_type = atoi(optarg);
            sndlib_name = EAS_GetDefaultSoundLibrary(sndlib_type);
            if ((sndlib_type < EAS_SNDLIB_WT) || (sndlib_type > EAS_SNDLIB_HYBRID) || (sndlib_name == NULL)) {
                fprintf(stderr, "invalid sound library: %d\n", sndlib_type);
                return EXIT_FAILURE;
            }
            break;
        }
        case '?':
            fprintf(stderr, "unknown option %c\n", optopt);
            return EXIT_FAILURE;
        default:
            return EXIT_FAILURE;
        }
    }

    ok = initializeLibrary();
    if (ok != EXIT_SUCCESS) {
        return ok;
    }

    for (index = optind; index < argc; index++) {
        ok = renderFile(argv[index]);
        if (ok != EXIT_SUCCESS) {
            break;
        }
    }

    shutdownLibrary();

    return ok;
}
